galsim/gsobject.py

```diff
@@ -1351,57 +1322,410 @@ def drawImage(self, image=None, nx=None, ny=None, bounds=None, scale=None, wcs=N
+            if n_photons == 0.: poisson_flux = True
+            else: poisson_flux = False
+
+        # Check that either n_photons is set to something or flux is set to something
+        if (n_photons == 0. and self.getFlux() == 1.
+            and area == 1. and exptime == 1.): # pragma: no cover
+            import warnings
+            warnings.warn(
+                    "Warning: drawImage for object with flux == 1, area == 1, and "
+                    "exptime == 1, but n_photons == 0.  This will only shoot a single photon.")
+
+        # Setup the rng if not provided one.
+        if rng is None:
+            ud = galsim.UniformDeviate()
+        elif isinstance(rng, galsim.BaseDeviate):
+            ud = galsim.UniformDeviate(rng)
+        else:
+            raise TypeError("The rng provided is not a BaseDeviate")
+
+        # Make sure the image is set up to have unit pixel scale and centered at 0,0.
+        if image.wcs != galsim.PixelScale(1.0):
+            raise ValueError("drawPhot requires an image with scale=1.0")
+        if image.center() != galsim.PositionI(0,0):
+            raise ValueError("drawPhot requires an image centered at 0,0")
+
+        Ntot, g = self._calculate_nphotons(n_photons, poisson_flux, max_extra_noise, ud)
+
+        if gain != 1.:
+            g /= gain
+
+        # total flux falling inside image bounds, this will be returned on exit.
+        added_flux = 0.
+
+        # Don't do more than this at a time to keep the  memory usage reasonable.
+        maxN = 100000
+
+        # Nleft is the number of photons remaining to shoot.
+        Nleft = Ntot
+        while Nleft > 0:
+            # Shoot at most maxN at a time
+            thisN = min(maxN, Nleft)
+
             try:
-                added_photons = prof.SBProfile.drawShoot(
-                    imview.image, n_photons, uniform_deviate, gain, max_extra_noise,
-                    poisson_flux, add_to_image)
-                image.added_flux = added_photons / area / exptime
+                phot_array = self.SBProfile.shoot(thisN, ud);
             except RuntimeError:  # pragma: no cover
                 # Give some extra explanation as a warning, then raise the original exception
                 # so the traceback shows as much detail as possible.
                 import warnings
                 warnings.warn(
-                    "Unable to draw this GSObject with method='phot'.  Perhaps it is a "+
+                    "Unable to draw this GSObject with photon shooting.  Perhaps it is a "+
                     "Deconvolve or is a compound including one or more Deconvolve objects.")
                 raise
-        else:
-            added_photons = prof.SBProfile.draw(imview.image, gain, wmult)
-            image.added_flux = added_photons / area / exptime
 
-        return image
+            phot_array.scaleFlux(g * thisN / Ntot);
+
+            added_flux += phot_array.addTo(image.image)
+            Nleft -= thisN;
```

Comment:
More that I failed to remove the semi-colons.  This was a big copy past from SBProfile.cpp.  Thanks for catching it.